go/database/mpt/proof_test.go

```diff
@@ -402,6 +384,276 @@ func TestWitnessProof_Extract_EmbeddedNode_In_Proof(t *testing.T) {
+			got, complete, err := proof.GetBalance(rootHash, address)
+			if !complete || err != nil {
+				return complete, err
+			}
+			if got, want := got, balance; got != want {
+				t.Errorf("unexpected balance: got %v, want %v", got, want)
+			}
+			return complete, err
+		}},
+		"GetCodeHash": {test: func(t *testing.T, proof WitnessProof, rootHash common.Hash, address common.Address, key common.Key) (bool, error) {
+			got, complete, err := proof.GetCodeHash(rootHash, address)
+			if !complete || err != nil {
+				return complete, err
+			}
+			if got, want := got, hash; got != want {
+				t.Errorf("unexpected code hash: got %v, want %v", got, want)
+			}
+			return complete, err
+		}},
+		"GetValue": {test: func(t *testing.T, proof WitnessProof, rootHash common.Hash, _ common.Address, key common.Key) (bool, error) {
+			got, complete, err := proof.GetState(rootHash, address, key)
+			if !complete || err != nil {
+				return complete, err
+			}
+			if got, want := got, value; got != want {
+				t.Errorf("unexpected value: got %v, want %v", got, want)
+			}
+			return complete, err
+		}},
+	}
+
+	desc := &Extension{
+		path: AddressToNibblePath(address, ctxt)[0:30],
+		next: &Account{address: address, pathLength: 34, info: info,
+			storage: &Extension{
+				path: KeyToNibblePath(key, ctxt)[0:40],
+				next: &Value{key: key, length: 24, value: value},
+			}},
+	}
+
+	root, node := ctxt.Build(desc)
+	totalProof := createReferenceProof(t, ctxt, &root, node)
+	rootHash, _ := ctxt.getHashFor(&root)
+
+	// corrupt non-root node in the proof
+	corruptedProof := WitnessProof{maps.Clone(totalProof.proofDb)}
+	for k := range corruptedProof.proofDb {
+		if k != rootHash {
+			corruptedProof.proofDb[k] = []byte{0xAA, 0xBB, 0xCC, 0xDD}
+		}
+	}
+
+	for name, test := range tests {
+		t.Run(fmt.Sprintf("%s", name), func(t *testing.T) {
+			complete, err := test.test(t, totalProof, rootHash, address, key)
+			if err != nil {
+				t.Fatalf("unexpected error: %v", err)
+			}
+			if !complete {
+				t.Fatalf("proof should be complete")
+			}
+		})
+		t.Run(fmt.Sprintf("%s invalid", name), func(t *testing.T) {
+			_, err := test.test(t, corruptedProof, rootHash, address, key)
+			if err == nil {
+				t.Fatalf("expected error")
+			}
+		})
+	}
+}
+
+func TestWitnessProof_Access_Proof_Fields_CompleteProofs_EmptyFields_AnotherAddress(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	address := common.Address{1}
+	differentAddress := common.Address{2}
+	key := common.Key{2}
+	info := AccountInfo{Nonce: common.Nonce{1}, Balance: common.Balance{1}, CodeHash: common.Hash{1}}
+
+	ctxt := newNodeContextWithConfig(t, ctrl, S5LiveConfig)
+
+	testGetters := map[string]struct {
+		test func(*testing.T, WitnessProof, common.Hash, common.Address, common.Key) (bool, error)
+	}{
+		"GetAccountInfo": {test: func(t *testing.T, proof WitnessProof, rootHash common.Hash, address common.Address, key common.Key) (bool, error) {
```

Comment:
I guess, as for the test above, the common parts of those `test` functions could be factored out.

Reply:
done